internal globals (such as NativeStyleMetrics) accessible in user code

Edit:  NativeStyleMetrics is marked as `is_internal`, and yet it can be looked up  in user code.
There is a check that we don't lookup internal elements, but that check is not active for globals.

### Original description


Slint 1.3.0 - 3deb620953f1307de1b9c0b5b757b5d8ed31a3a1

When trying to preview `left_side_panel.slint` project file in VScode I got this crash
```
thread 'main' panicked at 'Cannot find native global NativeStyleMetrics', internal/interpreter/global_component.rs:107:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[Info  - 9:38:33 PM] Connection to server got closed. Server will restart.
The Slint Language Server crashed. This is a bug.
Please open an issue on https://github.com/slint-ui/slint/issues
```

Compilation also fails
```
error[E0412]: cannot find type `NativeStyleMetrics` in this scope
     --> /home/rafal/Projekty/Rust/czkawka/target/debug/build/czkawka_slint-e3a803d91d1c663c/out/main_window.rs:27072:341
      |
27072 | ...Metrics : :: core :: pin :: Pin < sp :: Rc < r#NativeStyleMetrics >> , global_ColorPalette_257 : :: core :: pin :: Pin < sp :: Rc < In...
      |                                                 ^^^^^^^^^^^^^^^^^^^^ not found in this scope
      |
help: you might be missing a type parameter
      |
27071 |      # [allow (dead_code)] struct Globals_MainWindow<NativeStyleMetrics> {
      |                                                     ++++++++++++++++++++

```

Changes that caused problem - https://github.com/qarmin/czkawka/commit/f128dd6a881fa58fd3000762d8f8844222cf3100

I started to use global NativeStyleMetrics to get info if it is dark mode or not
```
export global ColorPalette {
    // Tabs at left side
    in-out property<color> tab_selected_color: NativeStyleMetrics.dark-color-scheme ? red : blue;
    in-out property<color> tab_hovered_color: NativeStyleMetrics.dark-color-scheme ? green : yellow;
}
```

--- a/internal/compiler/lookup.rs
+++ b/internal/compiler/lookup.rs
@@ -451,7 +451,6 @@ fn expression_from_reference(
 }
 
 /// Lookup for Globals and Enum.
-/// Note: for enums, the expression's value is `usize::MAX`
 struct LookupType;
 impl LookupObject for LookupType {
     fn for_each_entry<R>(
@@ -465,7 +464,7 @@ impl LookupObject for LookupType {
             }
         }
         for (name, ty) in ctx.type_register.all_elements() {
-            if let Some(r) = Self::from_element(ty).and_then(|e| f(&name, e)) {
+            if let Some(r) = Self::from_element(ty, ctx, &name).and_then(|e| f(&name, e)) {
                 return Some(r);
             }
         }
@@ -474,7 +473,7 @@ impl LookupObject for LookupType {
 
     fn lookup(&self, ctx: &LookupCtx, name: &str) -> Option<LookupResult> {
         Self::from_type(ctx.type_register.lookup(name))
-            .or_else(|| Self::from_element(ctx.type_register.lookup_element(name).ok()?))
+            .or_else(|| Self::from_element(ctx.type_register.lookup_element(name).ok()?, ctx, name))
     }
 }
 impl LookupType {
@@ -485,10 +484,20 @@ impl LookupType {
         }
     }
 
-    fn from_element(el: ElementType) -> Option<LookupResult> {
+    fn from_element(el: ElementType, ctx: &LookupCtx, name: &str) -> Option<LookupResult> {
         match el {
             ElementType::Component(c) if c.is_global() => {
-                Some(Expression::ElementReference(Rc::downgrade(&c.root_element)).into())
+                // Check if it is internal, but allow re-export (different name) eg: NativeStyleMetrics re-exported as StyleMetrics
+                if c.root_element
+                    .borrow()
+                    .builtin_type()
+                    .map_or(false, |x| x.is_internal && x.name == name)
+                    && !ctx.type_register.expose_internal_types
+                {
+                    None
+                } else {
+                    Some(Expression::ElementReference(Rc::downgrade(&c.root_element)).into())
+                }
             }
             _ => None,
         }
@@ -731,14 +740,18 @@ struct BuiltinNamespaceLookup;
 impl LookupObject for BuiltinNamespaceLookup {
     fn for_each_entry<R>(
         &self,
-        _ctx: &LookupCtx,
+        ctx: &LookupCtx,
         f: &mut impl FnMut(&str, LookupResult) -> Option<R>,
     ) -> Option<R> {
         None.or_else(|| f("Colors", LookupResult::Namespace(BuiltinNamespace::Colors)))
             .or_else(|| f("Math", LookupResult::Namespace(BuiltinNamespace::Math)))
             .or_else(|| f("Key", LookupResult::Namespace(BuiltinNamespace::Key)))
             .or_else(|| {
-                f("SlintInternal", LookupResult::Namespace(BuiltinNamespace::SlintInternal))
+                if ctx.type_register.expose_internal_types {
+                    f("SlintInternal", LookupResult::Namespace(BuiltinNamespace::SlintInternal))
+                } else {
+                    None
+                }
             })
     }
 }

--- a/internal/compiler/tests/syntax/lookup/global.slint
+++ b/internal/compiler/tests/syntax/lookup/global.slint
@@ -24,7 +24,10 @@ export X := Rectangle {
 //     ^error{Cannot access id 'SomeComp'}
 
     my_lowercase := Rectangle {
-
+        clip: NativeStyleMetrics.dark-color-scheme;
+        //    ^error{Cannot access id 'NativeStyleMetrics'}
+        visible: SlintInternal.dark-color-scheme;
+        //       ^error{Cannot access id 'SlintInternal'}
     }
 
     property <int> my_lowercase: 45;

--- a/internal/compiler/tests/syntax/lookup/property.slint
+++ b/internal/compiler/tests/syntax/lookup/property.slint
@@ -41,10 +41,15 @@ export X := Rectangle {
 //              ^error{Element 'Comp' does not have a property 'not_exist'}
     }
 
+    img := Image{}
+
     callback dummy;
     dummy => {
-        SlintInternal.dark-color-scheme = false;
+        ["hello"].length = 45;
 //      ^error{Assignment needs to be done on a property}
+        img.source.width = 45;
+//      ^error{Assignment needs to be done on a property}
+
     }
 
 }

--- a/tools/lsp/language/completion.rs
+++ b/tools/lsp/language/completion.rs
@@ -460,9 +460,7 @@ fn resolve_expression_scope(lookup_context: &LookupCtx) -> Option<Vec<Completion
     let mut r = Vec::new();
     let global = i_slint_compiler::lookup::global_lookup();
     global.for_each_entry(lookup_context, &mut |str, expr| -> Option<()> {
-        if str != "SlintInternal" {
-            r.push(completion_item_from_expression(str, expr));
-        }
+        r.push(completion_item_from_expression(str, expr));
         None
     });
     Some(r)
@@ -758,10 +756,16 @@ mod tests {
             res.iter().find(|ci| ci.label == "true").unwrap();
             res.iter().find(|ci| ci.label == "self").unwrap();
             res.iter().find(|ci| ci.label == "root").unwrap();
+            res.iter().find(|ci| ci.label == "TextInputInterface").unwrap();
 
             assert!(!res.iter().any(|ci| ci.label == "text"));
             assert!(!res.iter().any(|ci| ci.label == "red"));
             assert!(!res.iter().any(|ci| ci.label == "nope"));
+
+            assert!(!res.iter().any(|ci| ci.label == "Rectangle"));
+            assert!(!res.iter().any(|ci| ci.label == "Clip"));
+            assert!(!res.iter().any(|ci| ci.label == "NativeStyleMetrics"));
+            assert!(!res.iter().any(|ci| ci.label == "SlintInternal"));
         }
     }
 
@@ -860,6 +864,7 @@ mod tests {
         "#;
         let res = get_completions(source).unwrap();
         res.iter().find(|ci| ci.label == "LineEdit").unwrap();
+        res.iter().find(|ci| ci.label == "StyleMetrics").unwrap();
 
         let source = r#"
             import { Foo, 🔺} from "std-widgets.slint"

